---
title: "embryo_intestine_scRNAseq"
author: "Jessica Hill"
date: "9/22/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---

# Install packages
```{r}
#install.packages('Seurat')
#packageVersion('Seurat') # 5.0.1
#install.packages('Matrix')

#install.packages("hdf5r")
#BiocManager::install("biomaRt")
#install.packages("clustree")
#install.packages("tidyverse")
#install.packages("patchwork")
```

# Load packages
```{r echo=FALSE}
library(biomaRt)
library(Seurat)
library(tidyverse)
library(patchwork)
library(clustree)
```

---

# Load the data
```{r}
emb.GFPpooled.data <- Read10X_h5("./cellranger/all_GFP/outs/count/filtered_feature_bc_matrix.h5")
```

Get filternames from https://parasite.wormbase.org/biomart/martview
```{r}
mart <- useMart("parasite_mart", dataset = "wbps_gene", host = "https://parasite.wormbase.org", port = 443)

genes.df <- getBM(mart = mart, 
      filters = c("species_id_1010", "gene_name"),
      value = list("caelegprjna13758", rownames(emb.GFPpooled.data)),
      attributes = c('wbps_gene_id','external_gene_id', 'chromosome_name'))

#genes.df1 <- getBM(mart = mart, 
#      filters = c("species_id_1010", "wbps_gene_id"),
#      value = list("caelegprjna13758", rownames(emb.GFPpooled.data1)),
#      attributes = c('wbps_gene_id','external_gene_id', 'chromosome_name')
#      ) # total is 20061

mitoList = genes.df[genes.df$chromosome_name == 'MtDNA', 'external_gene_id']
genes.df$chromosome_name = NULL

c(nrow(genes.df), nrow(emb.GFPpooled.data))

# there are genes missing
nrow(emb.GFPpooled.data)-nrow(genes.df)
```

There are `r nrow(emb.GFPpooled.data)-nrow(genes.df)` genes that are now dead, remove them from the input matrix.

```{r}
emb.GFPpooled.data.filt <- emb.GFPpooled.data[rownames(emb.GFPpooled.data) %in% genes.df$external_gene_id,]

# The two data structures do not have the same order of WBGeneIDs
identical(rownames(emb.GFPpooled.data.filt), genes.df$external_gene_id)

# Use match to make sure they are in the same order
identical(rownames(emb.GFPpooled.data.filt),
          genes.df[match(rownames(emb.GFPpooled.data.filt), genes.df$external_gene_id),]$external_gene_id)

# now replace the WBGeneIDs with gene names
#rownames(emb.GFPpooled.data.filt) <- genes.df[
#  match(rownames(emb.GFPpooled.data.filt),
#        genes.df$wbps_gene_id)
#  ,]$external_gene_id

# are there NAs?
sum(!is.na(rownames(emb.GFPpooled.data.filt)))

head(emb.GFPpooled.data.filt)
```


# Set up the Seurat object
```{r}
emb.GFPpooled <- CreateSeuratObject(counts = emb.GFPpooled.data.filt, project = "emb_GFPpooled", min.cells = 3, min.features = 200)
```

```{r}
length(unique(genes.df$wbps_gene_id))
length(unique(rownames(emb.GFPpooled)))
rownames(emb.GFPpooled)[!(rownames(emb.GFPpooled) %in% genes.df$external_gene_id)]
```


# Quality control metrics
- `nFeature_RNA` is the number of genes detected in each cell. 
- `nCount_RNA` is the total number of molecules detected within a cell (AKA total UMIs per cell)
- Low `nFeature_RNA` indicates that the cell may be dead or dying or an empty droplet
- High `nCount_RNA` and/or `nFeature_RNA` indicates that the "cell" may in fact be a doublet (or multiplet)

```{r}
# show QC metrics for the first 5 cells 
head(emb.GFPpooled@meta.data, 5)
```

Mitochondrial genome percentage 
```{r}
mitoIndices = which(rownames(emb.GFPpooled) %in% mitoList)
emb.GFPpooled[["percent.mt"]] <- PercentageFeatureSet(emb.GFPpooled, features = mitoIndices)
```

Visualize QC metrics as a violin plot
```{r}
VlnPlot(emb.GFPpooled, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.VlnPlot.png")
```

FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything calculated by the object, i.e. columns in object metadata, PC scores etc.
```{r}
plot1 <- FeatureScatter(emb.GFPpooled, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_hline(yintercept = 50)
plot2 <- FeatureScatter(emb.GFPpooled, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.FeatureScatter.png")
```


Filter based on features.
cells being filtered due to mito percent: 
```{r}
sum(emb.GFPpooled$percent.mt > 50)/ length(emb.GFPpooled$percent.mt)

emb.GFPpooled <- subset(emb.GFPpooled, subset = nFeature_RNA > 200 & nFeature_RNA < 5000 & nCount_RNA < 50000 & percent.mt < 50)
```


Visualize after filtering
```{r}
VlnPlot(emb.GFPpooled, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.VlnPlot.filt.png")

FeatureScatter(emb.GFPpooled, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.FeatureScatter.filt.png")
```


# Normalize the data
After removing unwanted cells from the dataset, we next normalize the data. 
By default, we employ a global-scaling normalization method "LogNormalize" that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 is the default), and log-transforms the result. 

Normalized values are stored in emb.GFPpooled[["RNA]]@data

```{r}
emb.GFPpooled <- NormalizeData(emb.GFPpooled, normalization.method = "LogNormalize", scale.factor = 10000)
```


# Feature selection
Identify highly variable features to focus on in downstream analysis
```{r}
emb.GFPpooled <- FindVariableFeatures(emb.GFPpooled, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(emb.GFPpooled), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(emb.GFPpooled)
LabelPoints(plot = plot1, points = top10, repel = TRUE)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.VarFeatures.png")
```


# Scaling the data
Apply a linear transformation ("scaling") prior to dimensional reduction, PCA. 
Shifts the expression of each gene, so that the mean expression across cells is 0. 
Scales the expression of each gene, so that the variance across cells is 1. 
This gives equal weight in downstream analysis, so that highly expressed genes dont dominate. 
```{r}
all.genes <- rownames(emb.GFPpooled)
emb.GFPpooled <- ScaleData(emb.GFPpooled, features = all.genes)
```


# Linear dimensional reduction 
perform PCA on the scaled data. 
By default, only the previously determined vairable features are used as input. 
```{r}
emb.GFPpooled <- RunPCA(emb.GFPpooled, features = VariableFeatures(object = emb.GFPpooled))
```

Examine and visualize PCA results a few different ways
```{r}
print(emb.GFPpooled[["pca"]], dims = 1:5, nfeatures = 5)
```

```{r}
# top genes associated with reduction components
VizDimLoadings(emb.GFPpooled, dims = 1:2, reduction = "pca")
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.DimLoadings.png")
```

```{r}
DimPlot(emb.GFPpooled, reduction = "pca")
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.DimPlot.png")
```

```{r}
DimHeatmap(emb.GFPpooled, dims = 1, cells = 500, balanced = TRUE)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.HeatMap1.png")
```

```{r}
DimHeatmap(emb.GFPpooled, dims = 1:6, cells = 500, balanced = TRUE)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.HeatMap1-6.png")
```

# Determine dataset dimensionality
To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metafeature’ that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many components should we choose to include? 

Seurat implements a resampling test inspired by the JackStraw procedure. We randomly permute a subset of the data (1% by default) and rerun PCA, constructing a ‘null distribution’ of feature scores, and repeat this procedure. We identify ‘significant’ PCs as those who have a strong enrichment of low p-value features
```{r}
emb.GFPpooled <- JackStraw(emb.GFPpooled, num.replicate = 100)
emb.GFPpooled <- ScoreJackStraw(emb.GFPpooled, dims = 1:20)
```

The JackStrawPlot() function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). ‘Significant’ PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line). 
```{r}
JackStrawPlot(emb.GFPpooled, dims = 1:20)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.JackStrawPlot.png")
```


An alternative heuristic method generates an ‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each one.
```{r}
ElbowPlot(emb.GFPpooled)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ElbowPlot.png")
```

The elbow plot is helpful when determining how many PCs we need to capture the majority of the variation in the data. The elbow plot visualizes the standard deviation of each PC. Where the elbow appears is usually the threshold for identifying the majority of the variation. However, this method can be a bit subjective about where the elbow is located.


# Cluster the cells
Seurat applies a graph-based clustering approach.

Seurat first constructs a KNN graph based on the euclidean distance in PCA space, and refines the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors() function, and takes as input the previously defined dimensionality of the dataset.

To cluster the cells, Seurat then applies modularity optimization techniques such as the Louvain algorithm (default) or SLM, to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters() function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters.
```{r}
emb.GFPpooled <- FindNeighbors(emb.GFPpooled, dims = 1:8)

# deviation form original seurat pipeline
# generate a vector of different resolutions to see which one performs best 
emb.GFPpooled.clust <- FindClusters(
  object = emb.GFPpooled,
  reduction.type = "pca",
  resolution = c(0.2, 0.4, 0.8, 1, 1.2),
  dims.use = 1:8,
  save.SNN = TRUE)

## Now use Clustree to build a clustering tree based on the chosen resolutions specified above 
clustree(emb.GFPpooled.clust, prefix = "RNA_snn_res.")
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.Clustree.png")

# now choose the appropriate resolution for the given dimensions using the Clustree data
emb.GFPpooled <- FindClusters(emb.GFPpooled, resolution = 1.2)
```

# Run non-linear dimensional reduction (UMAP or tSNE)
Visualize and explore data with UMAP or tSNE.
The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP or tSNE, it is suggested to use the same PCs as input to the clustering analysis.
```{r}
emb.GFPpooled <- RunUMAP(emb.GFPpooled, dims = 1:8)
```

```{r}
# Note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
DimPlot(emb.GFPpooled, reduction = "umap", label = "TRUE", label.box = TRUE)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.UMAP.png")
```

# Find differentially expressed features (cluster biomarkers)
Find markers that define clusters via differential expression. By default, Seurat identifies positive and negative markers of a single cluster compared to all other cells. You can do for all clusters, you can test groups of clusters vs. each other, or against all cells. 

The min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. max.cells.per.ident can also be set and will downsample each identity class to have no more cells than whatever this is set to. 
```{r}
emb.GFPpooled.markers <- FindAllMarkers(emb.GFPpooled, only.pos = F, min.pct = 0.25, logfc.threshold = 1)

emb.GFPpooled.markers %>% 
  group_by(cluster) %>% filter(p_val_adj < .01) %>%
  slice_max(n = 2, order_by = avg_log2FC)


write.csv(emb.GFPpooled.markers,"./Seurat_output/GFP_pooled/emb_GFPpooled_markers.csv", row.names = TRUE)
```

```{r}
emb.GFPpooled.markers %>% group_by(cluster) %>% filter(p_val_adj < .01) %>% top_n(n=2, wt=avg_log2FC) -> top2

DoHeatmap(emb.GFPpooled, features = top2$gene,raster=F)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.HeatMap-top2.png")


FeaturePlot(emb.GFPpooled, features = c("elt-2","elt-7", 'ges-1'),raster=F)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.FeaturePlot-PostCtrls.png")


FeaturePlot(emb.GFPpooled, features = c("glh-1","chs-1", "myo-2", "myo-3", "hlh-2", "rab-3"), raster=F)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.FeaturePlot-NegCtrls.png")
```


# Remove contamination clusters (based on expression of elt-2, elt-7, and ges-1)
contamination clusters = 3, 8, 10, 11, 12

```{r}
emb.GFPpooled.specificClusters <- RunUMAP(subset(emb.GFPpooled, idents = c('3','8', '10', '11', '12'), invert = TRUE), dims = 1:8)

DimPlot(emb.GFPpooled.specificClusters, reduction = "umap", label = "TRUE", label.box = TRUE)
```

# now rerun PCA and re-lcuster 
```{r}
emb.GFPpooled.specificClusters <- RunPCA(emb.GFPpooled.specificClusters, features = VariableFeatures(object = emb.GFPpooled.specificClusters))

DimPlot(emb.GFPpooled.specificClusters, reduction = "pca")
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.DimPlot.png")

ElbowPlot(emb.GFPpooled.specificClusters)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.ElbowPlot.png")
```

```{r}
# Choose the number of PCs to use for re-clustering 
emb.GFPpooled.specificClusters <- FindNeighbors(emb.GFPpooled.specificClusters, dims = 1:8)

# check new clusters with Clustree
emb.GFPpooled.specificClusters.recluster <- FindClusters(
  object = emb.GFPpooled.specificClusters,
  reduction.type = "pca",
  resolution = c(0.2, 0.4, 0.8, 1, 1.2),
  dims.use = 1:8,
  save.SNN = TRUE)

clustree(emb.GFPpooled.specificClusters.recluster, prefix = "RNA_snn_res.")
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.Clustree.png")

emb.GFPpooled.specificClusters.recluster <- FindClusters(emb.GFPpooled.specificClusters, resolution = 1.2)

emb.GFPpooled.specificClusters.recluster <- RunUMAP(emb.GFPpooled.specificClusters.recluster, dims = 1:8)

DimPlot(emb.GFPpooled.specificClusters.recluster, reduction = "umap", label = "TRUE", label.box = TRUE, label.size = 9, pt.size = 4) + theme(legend.text = element_text(size = 20))
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.UMAP.png")
```

# Find new marker genes
```{r}
emb.GFPpooled.specificClusters.recluster.markers <- FindAllMarkers(emb.GFPpooled.specificClusters.recluster, only.pos = F, min.pct = 0.25, logfc.threshold = 1)

emb.GFPpooled.specificClusters.recluster.markers %>% 
  group_by(cluster) %>%
  slice_max(n = 2, order_by = avg_log2FC)

write.csv(emb.GFPpooled.specificClusters.recluster.markers,"./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.markers.csv", row.names = TRUE)
```

# Validate clusters 
```{r}
emb.GFPpooled.specificClusters.recluster.markers %>% group_by(cluster) %>% filter(p_val_adj < .01) %>% top_n(n=2, wt=avg_log2FC) -> top2_reclust


DoHeatmap(emb.GFPpooled.specificClusters.recluster, features = top2_reclust$gene,raster=F) + theme(text = element_text(size = 20))
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.recluster.HeatMap-top2.png")


# intestine marker genes
FeaturePlot(emb.GFPpooled.specificClusters.recluster, features = c("elt-2","elt-7", 'ges-1'),raster=F, ncol = 1)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.recluster.FeaturePlot-PostCtrls.png")


# Contamination marker genes: germline (glh-1, chs-1), muscle (myo-2, myo-3), neuron(hlh-2, rab-3)  
FeaturePlot(emb.GFPpooled.specificClusters.recluster, features = c("glh-1","chs-1", "myo-2", "myo-3", "hlh-2", "rab-3"), raster=F)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.recluster.FeaturePlot-NegCtrls.png")
```


# Perform gene ontology
```{r}
library(reticulate)
py_config()
# to modify the python to use in R, nano ~/.Renviron. just uncomment the right path and comment out the other. Will need to restart the Rsession before implimenting. 
use_condaenv("base")

tea <- import("tissue_enrichment_analysis")
tissue_df <- tea$fetch_dictionary("tissue")
phenotype_df <- tea$fetch_dictionary("phenotype")
go_df <- tea$fetch_dictionary("go")
```

```{r}
emb.GFPpooled.specificClusters.recluster.markers <- emb.GFPpooled.specificClusters.recluster.markers %>% left_join(genes.df, by = c("gene" = "external_gene_id")) 

emb.GFPpooled.specificClusters.recluster.markers <- emb.GFPpooled.specificClusters.recluster.markers %>% rename(WBGeneID = wbps_gene_id)
```

```{r}
clusters <- unique(emb.GFPpooled.specificClusters.recluster.markers$cluster)

my_tea_analysis <- function(input_df, tea_dict){
  tea_df <- data.frame()
  for (i in clusters){
  tea_set <- tea$enrichment_analysis(
    (input_df %>% 
       filter(cluster == i))$WBGeneID, 
    tissue_df = tea_dict 
    # %>% 
    #   filter(wbid %in% (input_df %>% 
    #                       filter(cluster %in% clusters))$WBGeneID)
  )
  if(nrow(tea_set) == 0){
    next
  } else {
  tea_df <- bind_rows(tea_df, data.frame(tea_set, cluster = i))
  }
}
tea_df
}

reorder_within <- function(x, by, within, fun = mean, sep = "___", ...) {
    new_x <- paste(x, within, sep = sep)
    stats::reorder(new_x, by, FUN = fun)
}

scale_x_reordered <- function(..., sep = "___") {
    reg <- paste0(sep, ".+$")
    ggplot2::scale_x_discrete(labels = function(x) gsub(reg, "", x), ...)
}

my_ontology_plot <- function(input_df, xlabel){
  input_df %>%
  filter(Q.value < 0.001,  Observed > 50) %>% 
  ggplot(aes(x = reorder_within(Term, -log10(Q.value), cluster), y = -log10(Q.value))) +
  geom_point(aes(size = Observed)) +
  coord_flip() +
  scale_x_reordered() +
  facet_grid(cluster~., scales = "free_y") +
  theme_bw() +
  xlab(xlabel) +
  guides(size = guide_legend(title = "# genes")) +
  scale_size_continuous(breaks = c(20,100,200))
}
```

```{r fig.height=15, fig.width=5}
my_ontology_plot(my_tea_analysis(emb.GFPpooled.specificClusters.recluster.markers %>% filter(p_val_adj < 0.05, avg_log2FC > 1), tissue_df), "Tissue Ontology")

ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.GO.png")
```

# Pull out new cluster marker genes 
```{r}
emb.GFPpooled.specificClusters.recluster.markers %>% group_by(cluster) %>% filter(p_val_adj < .01) %>% top_n(n=4, wt=avg_log2FC) -> top4

DoHeatmap(emb.GFPpooled.specificClusters.recluster, features = top4$gene,raster=F)

ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.HeatMap-top4.png")
```


# Pull out new cluster specific CLECs 
```{r}
emb.GFPpooled.specificClusters.recluster.markers %>% group_by(cluster) %>% filter(grepl("clec*",gene), p_val_adj < .01) %>% top_n(n=2, wt=avg_log2FC) -> top2clecs

DoHeatmap(emb.GFPpooled.specificClusters.recluster, features = top2clecs$gene,raster=F)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.HeatMap-top2clecs.png")
```


# Pull out cluster specific SPPs 
```{r fig.height=4, fig.width=7}
emb.GFPpooled.specificClusters.recluster.markers %>% group_by(cluster) %>% filter(grepl("^spp",gene), p_val_adj < .01) %>% top_n(n=10, wt=avg_log2FC) -> top10spps

DoHeatmap(emb.GFPpooled.specificClusters.recluster, features = top10spps$gene,raster=F)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.HeatMap-top10spp.png")
```

# Pull out cluster specific IRGs 
```{r fig.height=4, fig.width=7}
emb.GFPpooled.specificClusters.recluster.markers %>% group_by(cluster) %>% filter(grepl("^irg",gene), p_val_adj < .01) %>% top_n(n=10, wt=avg_log2FC) -> top10irgs

DoHeatmap(emb.GFPpooled.specificClusters.recluster, features = top10irgs$gene,raster=F)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.HeatMap-top10irg.png")
```



# plot genes of interest
From John Murray dataset 
    - intestine_anterior: ZC204.12, cpr-1, ceh-37
    - intestine_middle_and_posterior: irg-7, pal-1, cpr-1, ceh-37
    - intestine_far_posterior: irg-7, faah-1, pbo-4, psa-3

```{r}
# top marker gene for each cluster
FeaturePlot(emb.GFPpooled.specificClusters.recluster, pt.size = 0.1,
            features = (emb.GFPpooled.specificClusters.recluster.markers %>% group_by(cluster) %>% slice_max(n = 1, order_by = avg_log2FC))$gene)
```

```{r}
# genes from murray data
# intestine_anterior
FeaturePlot(emb.GFPpooled.specificClusters.recluster, features = c('ZC204.12', 'cpr-1', 'ceh-37'), raster=F, ncol = 2)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.Murray-Anterior.png") 

# intestine_middle_and_posterior
FeaturePlot(emb.GFPpooled.specificClusters.recluster, features = c('irg-7', 'pal-1', 'cpr-1', 'ceh-37'), raster=F, ncol = 2)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.Murray-Mid.png")

# intestine_far_posterior
FeaturePlot(emb.GFPpooled.specificClusters.recluster, features = c('faah-1', 'pbo-4', 'psa-3'), raster=F, ncol = 2)
ggsave("./Seurat_output/GFP_pooled/emb.GFPpooled.ContaminationRemoved.Murray-Posterior.png")
```



# Assign cell type identity to clusters 
```
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono",
    "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

